c_emulator/SoftFloat-3e/source/fall_reciprocal.c

```diff
@@ -0,0 +1,392 @@
+    bool sub = false;
+    switch(ret) {
+    case 0x001: // -inf
+    case 0x002: // -normal
+    case 0x004: // -subnormal
+    case 0x100: // sNaN
+        softfloat_exceptionFlags |= softfloat_flag_invalid;
+    case 0x200: //qNaN
+        uA.ui = defaultNaNF64UI;
+        break;
+    case 0x008: // -0
+        uA.ui = 0xfff0000000000000ul;
+        softfloat_exceptionFlags |= softfloat_flag_infinite;
+        break;
+    case 0x010: // +0
+        uA.ui = 0x7ff0000000000000ul;
+        softfloat_exceptionFlags |= softfloat_flag_infinite;
+        break;
+    case 0x080: //+inf
+        uA.ui = 0x0;
+        break;
+    case 0x020: //+ sub
+        sub = true;
+    default: // +num
+        uA.ui = rsqrte7(uA.ui, 11, 52, sub);
+        break;
+    }
+
+    return uA.f;
+}
+
+//user needs to truncate output to required length
+static inline uint64_t recip7(uint64_t val, int e, int s, int rm, bool sub,
+                              bool *round_abnormal)
+{
+    uint64_t exp = extract64(val, s, e);
+    uint64_t sig = extract64(val, 0, s);
+    uint64_t sign = extract64(val, s + e, 1);
+    const int p = 7;
+
+    static const uint8_t table[] = {
+        127, 125, 123, 121, 119, 117, 116, 114,
+        112, 110, 109, 107, 105, 104, 102, 100,
+        99, 97, 96, 94, 93, 91, 90, 88,
+        87, 85, 84, 83, 81, 80, 79, 77,
+        76, 75, 74, 72, 71, 70, 69, 68,
+        66, 65, 64, 63, 62, 61, 60, 59,
+        58, 57, 56, 55, 54, 53, 52, 51,
+        50, 49, 48, 47, 46, 45, 44, 43,
+        42, 41, 40, 40, 39, 38, 37, 36,
+        35, 35, 34, 33, 32, 31, 31, 30,
+        29, 28, 28, 27, 26, 25, 25, 24,
+        23, 23, 22, 21, 21, 20, 19, 19,
+        18, 17, 17, 16, 15, 15, 14, 14,
+        13, 12, 12, 11, 11, 10, 9, 9,
+        8, 8, 7, 7, 6, 5, 5, 4,
+        4, 3, 3, 2, 2, 1, 1, 0};
+
+    if (sub) {
+        while (extract64(sig, s - 1, 1) == 0)
+            exp--, sig <<= 1;
+
+        sig = (sig << 1) & make_mask64(0 ,s);
+
+        if (exp != 0 && exp != UINT64_MAX) {
+            *round_abnormal = true;
+            if (rm == 1 ||
+                (rm == 2 && !sign) ||
+                (rm == 3 && sign))
+                return ((sign << (s+e)) | make_mask64(s, e)) - 1;
+            else
+                return (sign << (s+e)) | make_mask64(s, e);
+        }
+    }
+
+    int idx = sig >> (s-p);
+    uint64_t out_sig = (uint64_t)(table[idx]) << (s-p);
+    uint64_t out_exp = 2 * make_mask64(0, e - 1) + ~exp;
+    if (out_exp == 0 || out_exp == UINT64_MAX) {
+        out_sig = (out_sig >> 1) | make_mask64(s - 1, 1);
+        if (out_exp == UINT64_MAX) {
+            out_sig >>= 1;
+            out_exp = 0;
+        }
+    }
+
+    return (sign << (s+e)) | (out_exp << s) | out_sig;
+}
+
+float16_t f16_recip7(float16_t in)
+{
+    union ui16_f16 uA;
+
+    uA.f = in;
+    unsigned int ret = f16_classify(in);
+    bool sub = false;
+    bool round_abnormal = false;
+    switch(ret) {
+    case 0x001: // -inf
```

Comment:
All the softfloat source files are 100% copied from the Spike softfloat directory. I treat these files as external library. Maybe it's not necessary to make modifications to these files. Since the complete floating-point framework will be rewritten in the future, I think the most efficient way at present is to keep these files and treat them as a temporarily used library, as their reliability has been verified in Spike. 

For floating-point functions in Sail, I think the best way is to have specific developers implementing them and have a uniform interface for different extensions to call them, because it will be used by many ISA extensions. For now most extensions still use softfloat functions and have a uniform interface in `riscv_softfloat_interface.sail`, so I'm trying to keep the same actions with them, instead of implementing my own version of floating-point functions in Sail, which will maybe mess up the current framework.

Reply:
The Sail model isn't Spike, and bad code existing in Spike doesn't justify it existing in the Sail model.

Reply:
Since the softfloat C implementation will be replaced in the future, it can be regarded as a temporary solution and the functional reliability is the most important. I think dealing with coding issues in the softfloat library is out of the scope of the vector PR.

Reply:
Anything you're adding in the vector PR is in scope for the vector PR

Reply:
At the beginning, I'm requested to use softfloat library in vector implementation, not maintaining softfloat code.

Reply:
Adding new code is more maintaining it that reusing what's already there. But regardless, it's up to the maintainers of the repository to decide what code goes in, not what you want.

Reply:
For example, these source files aren't even following the code style for SoftFloat, aside from the fact that you shouldn't need to add to the SoftFloat sources; I find it hard to believe you can't write the Sail implementation in terms of the existing primitives.

Reply:
At that time, I assumed it's better to keep all floating-point functions use the same softfloat interface instead of some use softfloat and some use native Sail functions. In the floating-point instruction PR, I mentioned where the new files come from, and there's no discussion about this issue, which makes me think it's an acceptable way. We're going to rewrite this part in pure Sail, but it will take a little longer time because we should rerun some tests about it.